### PR TITLE
Slice syntax for calibration groups

### PR DESCRIPTION
**Pull request by kbwestfall**
The main purpose of this is to allow for slice syntax for calibration groups in the pypeit file.  The calibration group can now be 'all' or be some combination of listed numbers and slices.  Let's say you have 10 calibration groups.  Here are some examples of what you can do and what it implies:

 - 'all' = '0,1,2,3,4,5,6,7,8,9'
 - ':4' = '0,1,2,3'
 - '3:5,8:' = '3,4,8,9'
 - '3,1:5,6' = '1,2,3,4,6'

Note in that last one that it's smart enough to only keep unique elements in the list.

Along the way, I also:
- Streamlined some of the instantiation at the beginning of `PypeIt.__init__`.
    - Moves the call to `default_pypeit_par` into `config_specific_par`.
    - Adds a `finalize_usr_build` function to `PypeItMetaData` to consolidate the few opaque steps when finishing the meta data build.

Some tests fail, but I'm hoping they will pass when @jhennawi's current PR is merged.  Dev suite passes on Shane Kast.  I'll rerun the tests and the full dev suite once I've merged in his PR.

Note that this branch currently includes the differences between master and develop, so this also will update develop to be consistent with master.

### CODE REVIEW

File: pypeit/metadata.py [hunk 1]
```diff
@@ -1,10 +1,10 @@
 """
 Provides a class that handles the fits metadata required by PypeIt.
 """
-from __future__ import print_function
-from __future__ import absolute_import
-from __future__ import division
-from __future__ import unicode_literals
+#from __future__ import print_function
```
Comment (profxj) [inline pypeit/metadata.py]: These can be removed

File: pypeit/metadata.py [hunk 2]
```diff
@@ -223,202 +228,88 @@ def get_meta_data_model():
...
-        Add the background-pair columns to the table.
-        
-        This include two columns called `comb_id` and `bkg_id` that
-        identify object and background frame pairs.  The string
-        indicates how these these columns should be added::
-
-            - `empty`: The columns are added but their values are all
-              originally set to -1.  **This is currently the only
-              option.**
+#    def _add_bkg_pairs(self, data, bkg_pairs):
```
Comment (profxj) [inline pypeit/metadata.py]: Just checking:  `calib_id` and `bkg_id` don't show up for 
multi-slit, right?  Unless requested.

File: pypeit/metadata.py [hunk 3]
```diff
@@ -510,44 +401,49 @@ def merge(self, usrdata, match_type=True):
...
-                else:
-                    in_time = date
-            ival = float(in_time) if self.spectrograph.timeunit == 'mjd' else in_time
-            tval = time.Time(ival, scale='tt', format=self.spectrograph.timeunit)
-            # Put MJD
-            return tval.mjd
-        
-        msgs.error('Bad time unit')
-    '''
+        self.get_frame_types(user=frametype)
```
Comment (profxj) [inline pypeit/metadata.py]: Thanks for cleaning this up!

File: pypeit/metadata.py [hunk 4]
```diff
@@ -1328,22 +1286,29 @@ def set_pypeit_cols(self, write_bkg_pairs=False):
...
-        self.table is modified in place
+        If none of the 'comb_id' values are other then -1, the
+        combination groups are set to be unique for each standard and
+        science frame.
+
+        :attr:`table` is edited in place.
 
         """
-        # Set comb_id
+        if 'comb_id' not in self.keys():
```
Comment (profxj) [inline pypeit/metadata.py]: Again, just checking that these tags don't show up in the .pypeit file
unless desired.

File: pypeit/pypeit.py [hunk 5]
```diff
@@ -74,47 +74,51 @@ def __init__(self, pypeit_file, verbosity=2, overwrite=True, reuse_masters=False
...
-        if 'setup' not in self.fitstbl.keys():
-            self.fitstbl['setup'] = setups[0]
-        self.fitstbl.get_frame_types(user=frametype)  # This sets them using the user inputs
-        self.fitstbl.set_defaults()  # Only does something if values not set in PypeIt file
-        self.fitstbl._set_calib_group_bits()
-        self.fitstbl._check_calib_groups()
-        # Write .calib file (For QA naming amongst other things)
+        #   - Interpret automated or user-provided data from the PypeIt
+        #   file
+        self.fitstbl.finalize_usr_build(frametype, setups[0])
```
Comment (profxj) [inline pypeit/pypeit.py]: Thanks for this.

File: pypeit/metadata.py (hunk 1 above)
Comment (kbwestfall) [reply pypeit/metadata.py]: Done.  And I just did all the rest of them, too...

File: pypeit/metadata.py (hunk 2 above)
Comment (kbwestfall) [reply pypeit/metadata.py]: Right, I didn't change `set_pypeit_cols`.  These columns are always added to `fitstbl` just like before, but only output if `write_bkg_pairs` is true.

File: pypeit/metadata.py (hunk 4 above)
Comment (kbwestfall) [reply pypeit/metadata.py]: Yep.

File: pypeit/pypeit.py (hunk 5 above)
Comment (kbwestfall) [reply pypeit/pypeit.py]: NP.  Hopefully the doc string for `finalize_usr_build` is clear as to what it's doing.

File: pypeit/metadata.py (hunk 3 above)
Comment (kbwestfall) [reply pypeit/metadata.py]: NP

**Comment by kbwestfall** [comment]
@profxj Thanks for the quick comments!  I moved the main parsing code to `core/parse.py` and added a test to `test_parse.py`.  I'll wait to merge until I can merge in @jhennawi's PR.

**Comment by kbwestfall** [comment]
Update: I've merged in both the current `master` and Joe's upcoming `cmartin_fixes` PR merge.  Tests pass.  I'll run the Keck dev-suite tonight.

**Comment by profxj** [comment]
Ready to merge this?

**Comment by kbwestfall** [comment]
Merging now.  Will update with master after coming PRs merge.

**Pull request merged**